lms/djangoapps/instructor/views/api.py

```diff
@@ -3592,3 +3596,151 @@ def _create_error_response(request, msg):
+    successful_usernames, invalid_usernames = [], []
+    for raw_user in usernames_or_expires:
+
+        user = User.objects.filter(
+            Q(email=raw_user['username_or_email']) | Q(username=raw_user['username_or_email'])
+        ).first()
+
+        if user:
+            banned_user = CourseEnrollmentBanned.objects.filter(email=user.email).first()
+            if banned_user:
+                banned_user.is_banned = False
+                banned_user.expiry = None
+                banned_user.save()
+                successful_usernames.append({'identifier': raw_user['username_or_email'], 'reason': ''})
+            else:
+                invalid_usernames.append(
+                    {'identifier': raw_user['username_or_email'], 'reason': 'User is not banned already'})
+        else:
+            invalid_usernames.append({'identifier': raw_user['username_or_email'],
+                                      'reason': 'username is not registered'})
+
+    return successful_usernames, invalid_usernames
+
+
+def _ban_username_or_email(course_id, usernames_or_expires):
+    successful_usernames, invalid_usernames = [], []
+    valid_email_re = '^[a-z0-9]+[\._]?[a-z0-9]+[@]\w+[.]\w+$'
+    expiry_invalid_states = ['past_time', 'invalid']
+
+    for raw_user in usernames_or_expires:
+
+        user = User.objects.filter(
+            Q(email=raw_user['username_or_email']) | Q(username=raw_user['username_or_email'])
+        ).first()
+
+        if raw_user['expiry'] in expiry_invalid_states:
+            reason = 'cannot set the time in past'
+
+            if raw_user['expiry'] == 'invalid':
+                reason = 'Invalid time format for the expiry due date'
+
+            invalid_usernames.append({'identifier': raw_user['username_or_email'], 'reason': reason})
+
+        elif user:
+            banned_user = CourseEnrollmentBanned.objects.filter(email=user.email).first()
+
+            if not banned_user:
+                banned_user = CourseEnrollmentBanned.objects.create(email=user.email, course_id=course_id)
+
+            banned_user.is_banned = True
+            banned_user.expiry = raw_user['expiry']
+            banned_user.save()
+            successful_usernames.append({'identifier': raw_user['username_or_email'], 'reason': ''})
+
+        elif re.search(valid_email_re, raw_user['username_or_email']):
+            banned_user = CourseEnrollmentBanned.objects.create(
+                email=raw_user['username_or_email'],
+                course_id=course_id,
+                expiry=raw_user['expiry']
+            )
+
+            banned_user.save()
+            successful_usernames.append({'identifier': raw_user['username_or_email'], 'reason': ''})
+        else:
+            invalid_usernames.append({'identifier': raw_user['username_or_email'], 'reason': 'username does not exist'})
+
+    return successful_usernames, invalid_usernames
+
+
+def _clean_username_and_expiry(user_and_expiry_raw):
```

Comment:
I think we can avoid this formating by refactoring the request data if the current one is not the client's requirement.